src/openloco/windows/stationwnd.cpp

```diff
@@ -62,12 +667,237 @@ namespace openloco::ui::windows::station
+
+        static void initEvents()
+        {
+            events.draw = draw;
+            events.on_mouse_up = on_mouse_up;
+            events.on_resize = on_resize;
+            events.on_update = common::update;
+            events.prepare_draw = prepare_draw;
+            events.text_input = common::text_input;
+            events.get_scroll_size = get_scroll_size;
+            events.tooltip = tooltip;
+            events.draw_scroll = draw_scroll;
+        }
+    }
+
+    namespace common
+    {
+        struct TabInformation
+        {
+            widget_t* widgets;
+            const widx widgetIndex;
+            window_event_list* events;
+            const uint64_t* enabledWidgets;
+        };
+
+        static TabInformation tabInformationByTabOffset[] = {
+            { station::widgets, widx::tab_station, &station::events, &station::enabledWidgets },
+            { cargo::widgets, widx::tab_cargo, &cargo::events, &cargo::enabledWidgets },
+            { cargo_ratings::widgets, widx::tab_cargo_ratings, &cargo_ratings::events, &common::enabledWidgets }
+        };
+
+        static void prepare_draw(window* self)
+        {
+            // Reset tab widgets if needed.
+            auto tabWidgets = tabInformationByTabOffset[self->current_tab].widgets;
+            if (self->widgets != tabWidgets)
+            {
+                self->widgets = tabWidgets;
+                self->init_scroll_widgets();
             }
+
+            // Activate the current tab.
+            self->activated_widgets &= ~((1 << widx::tab_station) | (1 << widx::tab_cargo) | (1 << widx::tab_cargo_ratings));
+            widx widgetIndex = tabInformationByTabOffset[self->current_tab].widgetIndex;
+            self->activated_widgets |= (1ULL << widgetIndex);
+
+            // Put station and town name in place.
+            auto station = stationmgr::get(self->number);
+            commonFormatArgs[0] = station->name;
+            commonFormatArgs[1] = station->town;
+            commonFormatArgs[2] = (station->flags & 15) + 230;
+
+            // Resize common widgets.
+            self->widgets[common::widx::frame].right = self->width - 1;
+            self->widgets[common::widx::frame].bottom = self->height - 1;
+
+            self->widgets[common::widx::caption].right = self->width - 2;
+
+            self->widgets[common::widx::close_button].left = self->width - 15;
+            self->widgets[common::widx::close_button].right = self->width - 3;
+
+            self->widgets[common::widx::panel].right = self->width - 1;
+            self->widgets[common::widx::panel].bottom = self->height - 1;
+        }
+
+        // 0x0048E5DF
+        static void text_input(window* self, widget_index callingWidget, char* input)
+        {
+            if (callingWidget != common::widx::caption)
+                return;
+
+            if (strlen(input) == 0)
+                return;
+
+            addr<0x009C68E8, string_id>() = string_ids::error_cant_rename_station;
```

Comment:
`static loco_global<string_id, 0x009C68E8> gGameCommandErrorTitle;` add this to your defines at the start of the file

Reply:
I should do this in the company window PR, too.